src/shared/udev-util.c

```diff
@@ -319,3 +322,50 @@ bool device_for_action(sd_device *dev, DeviceAction action) {
 
         return a == action;
 }
+
+int udev_rule_parse_value(char *str, char **ret_value, char **ret_endpos) {
+        char *value = str;
+        char *i, *j;
+        int r;
+        bool is_escaped;
+
+        /* value must be double quotated */
+        is_escaped = value[0] == 'e';
+        value += is_escaped;
+        if (value[0] != '"')
+                return -EINVAL;
+        value++;
+
+        if (!is_escaped) {
+                /* unescape double quotation '\"'->'"' */
+                for (i = j = value; *i != '"'; i++, j++) {
+                        if (*i == '\0')
+                                return -EINVAL;
+                        if (i[0] == '\\' && i[1] == '"')
+                                i++;
+                        *j = *i;
+                }
+                j[0] = '\0';
+        } else {
+                _cleanup_free_ char *unescaped = NULL;
+
+                /* find the end position of value */
+                for (i = value; *i != '"'; i++) {
+                        if (*i == '\0')
+                                return -EINVAL;
+                        if (i[0] == '\\' && i[1] != '\0')
+                                i++;
+                }
+                i[0] = '\0';
```

Comment:
I think this loop must be something like the following:
```
        bool slash = false;

        for (i = value; ;i++) {
                if (*i == '\0')
                        return -EINVAL;
                if (slash) {
                        slash = false;
                        continue;
                }
                if (*i == '"')
                        break;
                if (*i == '\\')
                        slash = true;
        }
        *i = '\0'
```
Please consider the following cases (slash is escaped):
```
e="\\ -> -EINVAL
e="\\" -> -EINVAL
e="\\"" -> \\"
e="\\\\" -> \\\\
e="\\\\\\" -> -EINVAL
e="\\\\\\"" -> \\\\\\"
(...)
```

Reply:
I think the two loops do the same thing:
![o](https://user-images.githubusercontent.com/4318379/97227181-140ef500-1810-11eb-9079-d07ce0f4e861.png)


<details>
<summary>clang -g -O1 -fsanitize=fuzzer,address fuzz.c -Wall</summary>

```c
#include <assert.h>
#include <errno.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>


int a(const char *value) {
        const char *i;
        for (i = value; *i != '"'; i++) {
                if (*i == '\0')
                        return -EINVAL;
                if (i[0] == '\\' && i[1] != '\0')
                        i++;
        }
        return i - value;
}

int a2(const char *value) {
        const char *i;
        for (i = value; *i != '"'; i++) {
                if (*i == '\\')
                        i++;
                if (*i == '\0')
                        return -EINVAL;
        }
        return i - value;
}

int b(const char *value) {
        const char *i;
        bool slash = false;

        for (i = value;; i++) {
                if (*i == '\0')
                        return -EINVAL;
                if (slash) {
                        slash = false;
                        continue;
                }
                if (*i == '"')
                        break;
                if (*i == '\\')
                        slash = true;
        }
        return i - value;
}

int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
        char *str;
        assert((str = malloc(size + 1)));
        memcpy(str, data, size);
        str[size] = 0;

        assert(a(str) == a2(str));
        assert(a(str) == b(str));
	free(str);
        return 0;
}
```
</details>

I've add some tests for the inputs you mentioned. And changed my loop to this (hopefully it is more clear):

```c

        for (i = value; *i != '"'; i++) {
                if (*i == '\\')
                        i++;
                if (*i == '\0')
                        return -EINVAL;
        }
```

Let me know if your version is preferrable.

Reply:
Ah, my bad...